### PR TITLE
feat: Enhanced "refine" keyword for wells.

### PR DESCRIPTION
**Pull request by nilscb**
-May be either a boolean or a number specifying number of new points inserted between pairs of existing points.

### CODE REVIEW

**Review by hkfb — COMMENTED** [review]
Would be nice with a story and some unit tests.

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts [hunk 1]
```diff
@@ -251,13 +251,19 @@ export default class WellsLayer extends CompositeLayer<WellsLayerProps> {
...
-                : data;
-
             const coarseData = coarsenWells(data);
 
+            const doRefine =
+                typeof this.props.refine === "number"
+                    ? (this.props.refine as number) > 1
+                    : (this.props.refine as boolean);
+
+            const noRefineSteps =
```
Comment (t0oF-azpn) [inline typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts:261]: noRefineSteps ? no prefix is misleading. You can replace by nbRefineSteps or imply refineSteps ;)

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts [hunk 2]
```diff
@@ -251,13 +251,19 @@ export default class WellsLayer extends CompositeLayer<WellsLayerProps> {
 
             removeDuplicates(data);
 
-            const refine = this.props.refine;
-            data = refine
-                ? splineRefine(data) // smooth well paths.
-                : data;
-
             const coarseData = coarsenWells(data);
```
Comment (t0oF-azpn) [inline typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts:254]: Not sure it's a good idea to refine on a coarsened well trajectory.
Why did you change the algorithm order ?

File: typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.ts [hunk 3]
```diff
@@ -177,12 +177,14 @@ export function CatmullRom(
  * in smoother curves with more points.
  * Assumes 3D data.
  */
-export function splineRefine(data_in: FeatureCollection): FeatureCollection {
+export function splineRefine(
+    data_in: FeatureCollection,
+    noSteps = 5
```
Comment (t0oF-azpn) [inline typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.ts:182]: noSteps means you want no steps ? ;)
steps, stepCount or nbSteps is better...

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts (hunk 2 above)
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts:254]: corasening is done first so that the refinements are done on the on the original data. coarsenWells returns a coaresened verion of the data but does not chenge the original data.

Comment (t0oF-azpn) [reply typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts:254]: OK, thank you

**Review by t0oF-azpn — APPROVED** [review]

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts [hunk 4]
```diff
@@ -90,7 +90,7 @@ export interface WellsLayerProps extends ExtendedLayerProps {
     logrunName: string;
     logRadius: number;
     logCurves: boolean;
-    refine: boolean;
+    refine: boolean | number;
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts:93]: I feel that this needs documentation, especially now that it can be a number

File: typescript/packages/subsurface-viewer/src/layers/wells/utils/generateSyntheticWell.ts [hunk 5]
```diff
@@ -17,12 +17,12 @@ export function generateSynteticWell() {
                                 [0, 0, 0],
                                 [0, 0, -100],
                                 [0, 0, -200],
-                                [0, 0, -200],
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/wells/utils/generateSyntheticWell.ts:20]: You made this earlier to verify well trajectories with duplicate consecutive surveys. Suggest that you create a new example well for the new test. Also, the only thing this function does is return a variable, so might as well just be a variable.

**Review by hkfb — CHANGES_REQUESTED** [review]

File: typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.test.ts [hunk 6]
```diff
@@ -44,4 +82,27 @@ describe("remove duplicates", () => {
             removeConsecutiveDuplicates(coordsEmpty, mdsEmpty)[1]
         ).toStrictEqual(mdsEmpty);
     });
+
+    // Test splineRefine functions
+    it("should not refine if given invalid input", () => {
+        expect(
+            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+            // @ts-ignore
+            splineRefine(testWell as unknown as FeatureCollection, 0).features[0].geometry.geometries[1].coordinates.length // eslint-disable-line
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.test.ts:91]: Could you define the `testWell` const as `FeatureCollection` instead of casting it here?

**Review by hkfb — CHANGES_REQUESTED** [review]

**Review by hkfb — CHANGES_REQUESTED** [review]
Still missing a story for the `refine` parameter

**Review by t0oF-azpn — CHANGES_REQUESTED** [review]

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts [hunk 7]
```diff
@@ -90,7 +90,12 @@ export interface WellsLayerProps extends ExtendedLayerProps {
     logrunName: string;
     logRadius: number;
     logCurves: boolean;
-    refine: boolean;
+    /** If true will refine the well path by interpolating it using a spline and super sample it
+     * so that one by default gets 5 new vertices along the path between each pair of original vertices.
```
[the remark this answers is not in the document]
Comment (t0oF-azpn) [reply typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts:94]: Can you fix this doc, please ?

[the remark this answers is not in the document]
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts:94]: Done

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts [hunk 8]
```diff
@@ -90,7 +90,12 @@ export interface WellsLayerProps extends ExtendedLayerProps {
...
+     * May also be specified as a number greater than 1 in which case the number of new vertices
```
[the remark this answers is not in the document]
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts:95]: Done

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (t0oF-azpn) [reply typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts:94]: No, it's not

**Comment by hkfb** [comment]
I think it would be nice with a story that demonstrates the effect on a curved well trajectory with a low sample rate, ie. to be able to see the effect of the smoothing.

**Review by t0oF-azpn — APPROVED** [review]

**Review by t0oF-azpn — APPROVED** [review]

**Review by t0oF-azpn — APPROVED** [review]

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx [hunk 9]
```diff
@@ -415,6 +454,122 @@ AllWellHeadsHidden.parameters = {
...
+const BBox = [-100, -100, -250, 100, 100, 0] as [
+    number,
+    number,
+    number,
+    number,
+    number,
+    number,
+];
+
+export const WellsRefine: StoryFn<typeof SubsurfaceViewer> = (args) => {
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx:501]: I can't see the story, is it just me?

image

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx [hunk 10]
```diff
@@ -415,6 +454,122 @@ AllWellHeadsHidden.parameters = {
...
+    const [refineNumber, setRefineNumber] = React.useState<number>(1);
+
+    const props = {
+        ...args,
+        layers: [
+            {
+                "@@type": "WellsLayer",
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx:508]: Should use the class syntax in order to get typing

**Review by hkfb — CHANGES_REQUESTED** [review]

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx (hunk 10 above)
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx:508]: For some reason it doe not work witn "new". It does not pick up the input data changes. I suggest maybe we could do it this way fo rnow and later investigate further why "new" does not always work,

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx (hunk 9 above)
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx:501]: It should be there.

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx [hunk 11]
```diff
@@ -415,6 +454,122 @@ AllWellHeadsHidden.parameters = {
...
+            </div>
+            <Slider
+                min={1}
+                max={10}
+                defaultValue={1}
+                step={1}
+                onChange={handleChange}
+                valueLabelDisplay={"auto"}
+            />
+        </Root>
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx:540]: There's a lot of wasted vertical space between the story and the controls, but I guess it does the job

image

**Review by hkfb — APPROVED** [review]
